### PR TITLE
[sentencepiece] Adds Windows support

### PR DESCRIPTION
**Pull request by Lundez**
## Description ##

Brief description of what this PR is about

- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

**Comment by frankfliu** [comment]
@Lundez Thanks for contribution. You should be able to test in your folk (must in master branch).

You can ignore upload to S3 part when running in your folk.

I'm not sure if we need any CMake changes for this PR.

File: .github/workflows/native_s3_sentencepiece.yml [hunk 1]
```diff
@@ -71,3 +71,35 @@ jobs:
...
+          ./gradlew :extensions:sentencepiece:compileJNI
+          ./gradlew -Pjni :extensions:sentencepiece:test
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-east-2
+      - name: Copy files to S3 with the AWS CLI
+        run: |
```
Comment (frankfliu) [inline .github/workflows/native_s3_sentencepiece.yml:101]: shell: bash

The command requires gitbash

**Comment by Lundez** [comment]
> @Lundez Thanks for contribution. You should be able to test in your folk (must in master branch).
> 
> You can ignore upload to S3 part when running in your folk.
> 
> I think there are few piece is missing:
> 
> 1. The build.gradle also need to be update: https://github.com/deepjavalibrary/djl/blob/master/extensions/sentencepiece/build.gradle#L64-L68
> 2. We need add a build.cmd file for windows
> 3. I'm not sure if we need any CMake changes for this PR.

Ok, I'll take a look further. This is simply a draft PR for now :) 
Most likely I'll fix it during the weekend!

File: extensions/sentencepiece/build.gradle [hunk 2]
```diff
@@ -62,7 +62,8 @@ apply from: file("${rootProject.projectDir}/tools/gradle/cpp-formatter.gradle")
 task compileJNI {
     doFirst {
         if (System.properties['os.name'].toLowerCase(Locale.ROOT).contains("mac")
-                || System.properties['os.name'].toLowerCase(Locale.ROOT).contains("linux")) {
+                || System.properties['os.name'].toLowerCase(Locale.ROOT).contains("linux")
+                || System.properties['os.name'].toLowerCase(Locale.ROOT).contains("windows")) {
```
Comment (frankfliu) [inline extensions/sentencepiece/build.gradle:66]: I don't think this will work for windows, you need specify full "git-bash.exe" command line, see: https://stackoverflow.com/questions/57614140/execute-git-bash-from-gradle-task

Or you can create a "build.cmd" file for windows: https://github.com/deepjavalibrary/djl/blob/master/engines/pytorch/pytorch-native/build.gradle#L31-L33

File: extensions/sentencepiece/build.sh [hunk 3]
```diff
@@ -39,4 +39,7 @@ if [[ $PLATFORM == 'darwin' ]]; then
 elif [[ $PLATFORM == 'linux' ]]; then
   mkdir -p build/jnilib/linux-x86_64
   cp -f build/libsentencepiece_native.so build/jnilib/linux-x86_64/
+elif [[ $PLATFORM == 'windows' ]]; then
```
Comment (frankfliu) [inline extensions/sentencepiece/build.sh:42]:
```
if [ "$(expr substr $PLATFORM 1 5)" == "mingw" ]; then
```

File: extensions/sentencepiece/build.gradle (hunk 2 above)
Comment (frankfliu) [inline extensions/sentencepiece/build.gradle:66]: I don't think this work for windows, you need specify a full "git-bash.exe" path. see: https://stackoverflow.com/questions/57614140/execute-git-bash-from-gradle-task


or you can create a "build.cmd" file, see: https://github.com/deepjavalibrary/djl/blob/master/engines/pytorch/pytorch-native/build.gradle#L31-L33

**Review by frankfliu — COMMENTED** [review]
cmake doesn't support -j option, you need to the following:

```
if [ "$(expr substr $PLATFORM 1 5)" == "mingw" ]; then   
  cmake --build . --config Release
 else
   cmake --build . --config Release -- -j "${NUM_PROC}" 
fi 

```

File: extensions/sentencepiece/build.sh [hunk 4]
```diff
@@ -39,4 +39,7 @@ if [[ $PLATFORM == 'darwin' ]]; then
...
+  mkdir -p build/jnilib/windows-x86_64
+  cp -f build/libsentencepiece_native.so build/jnilib/windows-x86_64/
```
Comment (frankfliu) [inline extensions/sentencepiece/build.sh:44]:
```suggestion
  cp -f build/sentencepiece_native.dll build/jnilib/windows-x86_64/
```

File: extensions/sentencepiece/build.sh [hunk 5]
```diff
@@ -39,4 +39,7 @@ if [[ $PLATFORM == 'darwin' ]]; then
 elif [[ $PLATFORM == 'linux' ]]; then
   mkdir -p build/jnilib/linux-x86_64
   cp -f build/libsentencepiece_native.so build/jnilib/linux-x86_64/
+elif [[ $PLATFORM == 'windows' ]]; then
+  mkdir -p build/jnilib/windows-x86_64
```
Comment (frankfliu) [inline extensions/sentencepiece/build.sh:43]:
```suggestion
  mkdir -p build/jnilib/win-x86_64
```

File: extensions/sentencepiece/build.gradle (hunk 2 above)
Comment (Lundez) [reply extensions/sentencepiece/build.gradle]: Trying a cmd file

**Comment by Lundez** [comment]
@frankfliu https://github.com/Lundez/djl/actions/runs/1569820490 seems to run :)
Gonna remove AWS comments again

**Comment by frankfliu** [comment]
@Lundez 
I made a few fixed on top your PR, but it seems still broken. On Windows, the `.dll` file is not generated.
I believe CMakeList.txt file need to be updated for Windows, see: https://github.com/levyfan/sentencepiece-jni/blob/master/src/main/native/CMakeLists.txt#L50-L56, there might be some other flags needed.

**Comment by Lundez** [comment]
> @Lundez 
> I made a few fixed on top your PR, but it seems still broken. On Windows, the `.dll` file is not generated.
> I believe CMakeList.txt file need to be updated for Windows, see: https://github.com/levyfan/sentencepiece-jni/blob/master/src/main/native/CMakeLists.txt#L50-L56, there might be some other flags needed.

I'll take a look tomorrow. 👍

**Comment by Lundez** [comment]
Don't have windows myself, so hard to test this. I'll give it some more time tomorrow, been very busy today.

**Comment by Lundez** [comment]
@frankfliu it seems to be created based on the logs: https://pipelines.actions.githubusercontent.com/vaIqj4dCv5ckseiiqJ6l5KKSICSYkSf0qDCf97KwQ9WmgmyDyY/_apis/pipelines/1/runs/5/signedlogcontent/4?urlExpires=2021-12-14T16%3A24%3A23.3573049Z&urlSigningMethod=HMACV1&urlSignature=nD0kwdqTjE8qtFLbjNQgkDQw8DXgtDIcZKlY%2FmFM7uU%3D

> 2021-12-14T16:19:15.1459755Z   sentencepiece_native.vcxproj -> D:\a\djl\djl\extensions\sentencepiece\build\Release\sentencepiece_native.dll

> 2021-12-14T16:20:38.1959663Z     [Test worker] INFO ai.djl.sentencepiece.jni.LibUtils - Extracting /native/lib/win-x86_64/sentencepiece_native.dll to cache ...

though it fails on 
> Gradle suite > Gradle test > ai.djl.sentencepiece.SpTextEmbeddingTest > testEmbedding FAILED 2021-12-14T16:20:38.1970138Z     java.lang.IllegalStateException at SpTextEmbeddingTest.java:39

I'm not sure how I debug in GitHub Actions. And as I don't have Windows it's hard to run locally.. Ideas? Or is the DLL still missing you think?

**Comment by frankfliu** [comment]
@Lundez No worry, we will look into this.

@stu1130 can you review CMake changes for windows?

**Comment by Lundez** [comment]
@frankfliu got more logging on (`--info`)

> Gradle suite > Gradle test > ai.djl.sentencepiece.SpVocabularyTest > testTokenIdConversion FAILED
    java.lang.IllegalStateException: SentencePiece library not found: /native/lib/win-x86_64/sentencepiece_native.dll
        at ai.djl.sentencepiece.jni.LibUtils.copyJniLibraryFromClasspath(LibUtils.java:66)

**Review by frankfliu — APPROVED** [review]
